sockeye/dump_decoder_states.py

```diff
@@ -0,0 +1,239 @@
+            )
+        else:
+            start = self.tail_idx
+            end = self.tail_idx + num_entries
+            self.mmap[start:end] = array
+
+            self.tail_idx += num_entries
+
+
+class StateDumper:
+
+    def __init__(self,
+                 model: SockeyeModel,
+                 source_vocabs: List[Vocab],
+                 target_vocabs: List[Vocab],
+                 dump_prefix: str,
+                 max_seq_len_source: int,
+                 max_seq_len_target: int,
+                 state_data_type: str,
+                 word_data_type: str,
+                 device: pt.device) -> None:
+        self.model = model
+        self.source_vocabs = source_vocabs
+        self.target_vocabs = target_vocabs
+        self.device = device
+        self.traced_model = None
+        self.max_seq_len_source = max_seq_len_source
+        self.max_seq_len_target = max_seq_len_target
+
+        self.dump_prefix = dump_prefix
+        self.state_dump_file = None
+        self.words_dump_file = None
+
+        # info for KNNConfig
+        self.dump_size = 0
+        self.dimension = None
+        self.state_data_type = utils.get_numpy_dtype(state_data_type)
+        self.word_data_type = utils.get_numpy_dtype(word_data_type)
+
+    @staticmethod
+    def probe_token_count(target: str, max_seq_len: int) -> int:
+        token_count = 0
+        with codecs.open(target, 'r', 'utf-8') as f:
+            for line in f:
+                token_count += min(len(line.strip().split(' ')) + 1, max_seq_len)  # +1 for EOS
+        return token_count
+
+    def init_dump_file(self, initial_size: int) -> None:
+        self.dimension = self.model.config.config_decoder.model_size
+
+        self.state_dump_file = NumpyMemmapStorage(get_state_dump_filename(self.dump_prefix),
+                                                  self.dimension, self.state_data_type)
+        self.words_dump_file = NumpyMemmapStorage(get_word_dump_filename(self.dump_prefix),
+                                                  1, self.word_data_type)
+        self.state_dump_file.open(initial_size, 1)
+        self.words_dump_file.open(initial_size, 1)
+
+    def build_states_and_dump(self,
+                              sources: List[str],
+                              targets: List[str],
+                              batch_size: int) -> None:
+        assert self.state_dump_file != None, \
+               "You should call probe_token_count first to initialize the dump files."
+
+        # get data iter
+        data_iter = data_io.get_scoring_data_iters(
+            sources=sources,
+            targets=targets,
+            source_vocabs=self.source_vocabs,
+            target_vocabs=self.target_vocabs,
+            batch_size=batch_size,
+            max_seq_len_source=self.max_seq_len_source,
+            max_seq_len_target=self.max_seq_len_target
+        )
+
+        for batch_no, batch in enumerate(data_iter, 1):
```

Comment:
You could consider wrapping this in a `with pt.inference_mode()` (also see [here](https://github.com/awslabs/sockeye/blob/fbbf360303b0db1e9160c9eb62ce24102ca8e772/sockeye/inference.py#L941))